platforms/android/src/inject.rs

```diff
@@ -126,218 +95,200 @@ fn inner_adapter_from_handle(handle: jlong) -> Option<Arc<Mutex<InnerInjectingAd
 }
 
-extern "system" fn set_text_selection(
-    mut env: JNIEnv,
-    class: JClass,
+extern "system" fn perform_action<'local>(
+    mut env: JNIEnv<'local>,
+    _class: JClass<'local>,
     adapter_handle: jlong,
-    host: JObject,
+    host: JObject<'local>,
     virtual_view_id: jint,
-    anchor: jint,
-    focus: jint,
+    action: jint,
+    arguments: JObject<'local>,
 ) -> jboolean {
-    let Some(inner_adapter) = inner_adapter_from_handle(adapter_handle) else {
+    let Some(action) = PlatformAction::from_java(&mut env, action, &arguments) else {
         return JNI_FALSE;
     };
-    let mut inner_adapter = inner_adapter.lock().unwrap();
-    if let Some(events) = inner_adapter.set_text_selection(virtual_view_id, anchor, focus) {
-        events.raise(&mut env, &class, &host);
-        JNI_TRUE
-    } else {
-        JNI_FALSE
-    }
-}
-
-extern "system" fn collapse_text_selection(
-    mut env: JNIEnv,
-    class: JClass,
-    adapter_handle: jlong,
-    host: JObject,
-    virtual_view_id: jint,
-) -> jboolean {
     let Some(inner_adapter) = inner_adapter_from_handle(adapter_handle) else {
         return JNI_FALSE;
     };
     let mut inner_adapter = inner_adapter.lock().unwrap();
-    if let Some(events) = inner_adapter.collapse_text_selection(virtual_view_id) {
-        events.raise(&mut env, &class, &host);
-        JNI_TRUE
-    } else {
-        JNI_FALSE
-    }
+    let Some(events) = inner_adapter.perform_action(virtual_view_id, &action) else {
+        return JNI_FALSE;
+    };
+    drop(inner_adapter);
+    events.raise(&mut env, &host);
+    JNI_TRUE
 }
 
-extern "system" fn traverse_text(
-    mut env: JNIEnv,
-    class: JClass,
+extern "system" fn on_hover_event<'local>(
+    mut env: JNIEnv<'local>,
+    _class: JClass<'local>,
     adapter_handle: jlong,
-    host: JObject,
-    virtual_view_id: jint,
-    granularity: jint,
-    forward: jboolean,
-    extend_selection: jboolean,
+    host: JObject<'local>,
+    action: jint,
+    x: jfloat,
+    y: jfloat,
 ) -> jboolean {
     let Some(inner_adapter) = inner_adapter_from_handle(adapter_handle) else {
         return JNI_FALSE;
     };
     let mut inner_adapter = inner_adapter.lock().unwrap();
-    if let Some(events) = inner_adapter.traverse_text(
-        virtual_view_id,
-        granularity,
-        forward == JNI_TRUE,
-        extend_selection == JNI_TRUE,
-    ) {
-        events.raise(&mut env, &class, &host);
-        JNI_TRUE
-    } else {
-        JNI_FALSE
-    }
+    let Some(events) = inner_adapter.on_hover_event(action, x, y) else {
+        return JNI_FALSE;
+    };
+    drop(inner_adapter);
+    events.raise(&mut env, &host);
+    JNI_TRUE
 }
 
-fn delegate_class(env: &mut JNIEnv) -> Result<&'static JClass<'static>> {
+fn delegate_class(env: &mut JNIEnv) -> &'static JClass<'static> {
     static CLASS: OnceCell<GlobalRef> = OnceCell::new();
-    let global = CLASS.get_or_try_init(|| {
+    let global = CLASS.get_or_init(|| {
```

Comment:
You could use a `LazyCell` here right? Or maybe even better, use `std::sync::OnceLock` until our MSRV permit us to use `LazyLock`, which would allow us to drop our dependency on `once_cell`.